### DISCUSSION TITLE
Physics-related: a electromagnetic buzzer that receives two or more PWM waves of different frequencies

### DISCUSSION THREAD

**Discussion by Wind-stormger** [micropython/micropython#12273] [Hardware & Peripherals] [answered]
1. Can a electromagnetic buzzer emit sounds of all frequencies after receiving **two** or **more** PWM waves of different frequencies?
2. Is the frequency it emits really the frequency we set?  I don't have the equipment to measure it.
3. At different frequencies, it sounds very different in loudness. At least I think so.

**Comment by jimmo** [comment, marked as answer]
> Can a electromagnetic buzzer emit sounds of all frequencies after receiving two or more PWM waves of different frequencies?

Not quite sure what you mean here. Do you mean:
a) You are generating a signal made of some combination of two PWM signals at different frequencies?
b) You generate a signal at one frequency, then another at another frequency?

> Is the frequency it emits really the frequency we set?

Yes. If you have a mobile phone then you do indeed have the equipment! Have a look at for example https://phyphox.org/ -- it includes some good tools for doing analysis with the phone's microphone.

> At different frequencies, it sounds very different in loudness. At least I think so.

Yes, absolutely. Look up the datasheet for your buzzer, it will have a frequency response curve.

**Comment by Wind-stormger** [comment]
> Not quite sure what you mean here. Do you mean:
> a) You are generating a signal made of some combination of two PWM signals at different frequencies?
> b) You generate a signal at one frequency, then another at another frequency?

Two pins, respectively generate two PWMs with different frequencies, output to the same buzzer.

**Comment by robert-hh** [comment]
You must not connect two output pins together. If you want to PWM pins drive one buzzer, you have to decouple the outputs, e.g. with diodes and/or resistors. And, as Jimmo told, a simple buzzer is like a speaker, but with a very small frequency range. There are also buzzers which just need power to generate a tone. They have an internal frequency generation circuit. If you apply a (low) frequency PWM to these, you may get some strange noise.

**Reply by Wind-stormger** [reply]
Thank you, is there any reference circuit design for decouple the outputs?

**Reply by Wind-stormger** [reply]
In particular, I would like to input multiple PWM waves to a buzzer.

**Reply by jimmo** [reply]
The circuit you're looking for is called an "audio mixer".

Robert makes a good point -- you said you had a "buzzer", and typically (to me) that would be something that you apply a DC voltage to and it generates its own frequency. Like he says, if you do any for of PWM to this, it will not work.

I assumed because you were talking about PWM that you do in fact have a "speaker" (and probably a piezoelectric one given the indication of its limited frequency resonse).

**Reply by Wind-stormger** [reply]
I did mean the electromagnetic buzzer.

I have an electromagnetic buzzer,  not piezoelectric, I can clearly identify the sounds it makes at different frequencies.

However, a dynamic speaker does produce a more recognizable sound at different PWM frequencies.

**Comment by sosi-deadeye** [comment]
> 1. Can a electromagnetic buzzer emit sounds of all frequencies after receiving **two** or **more** PWM waves of different frequencies?
> 
>     2. Is the frequency it emits really the frequency we set?  I don't have the equipment to measure it.
> 
>     3. At different frequencies, it sounds very different in loudness. At least I think so.

1.) Yes, but you have to mix those signals and prevent the feedback of the signal back to the outputs.
2.) In theory, you then have two frequencies mixed. Since only the sine occurs in nature, the signal produces the two desired frequencies and harmonics.

The harmonics are, for example, the whirring of electric drives.

3.) We don't hear all frequencies the same "loudness". I think it's called psychoacoustic.

**Comment by Wind-stormger** [comment]
I tried searching for "audio mixer" and found it to be extremely expensive.

I thought, why don't I use one more buzzer, hahaha. 😄 Multichannel is cheaper than "audio mixer".

Thank you all. @sosi-deadeye @jimmo @robert-hh @anotherjesse

**Reply by Wind-stormger** [reply]
Unless there is some economical way to DIY an "audio mixer"~~

**Reply by robert-hh** [reply]
You can use that simple circuit. It protects the board. The value of R depends on the impedance of the buzzer, but is low, like 10 - 33 Ohm.

![buzzer](https://github.com/micropython/micropython/assets/12476868/00f66c0a-b3bd-49ad-898e-e15dc78296e5)

**Reply by Wind-stormger** [reply]
Thanks, I'll go buy two 1N4148's and try it.

**Comment by scruss** [comment]
Alternatively, you can achieve a polyphonic sound effect from a single PWM output using square-wave frequencies overlaid on a very fast PWM 'carrier' signal (for example: [4000AD | PROTODOME](https://protodome.bandcamp.com/album/4000ad), generated on an ATMega microcontroller feeding a single pin). Whether we have the hardware facilities to do this on various MicroPython platforms, I don't know, as it requires extremely precise PWM control.

**Reply by Wind-stormger** [reply]
Very amazing, it is a method that has extremely high requirements for execution speed and accuracy!
